qualtran/bloqs/mod_arithmetic/mod_division.py

```diff
@@ -0,0 +1,660 @@
+    """The single full iteration of Kaliski. see figure 15 of https://arxiv.org/pdf/2302.06639."""
+
+    bitsize: 'SymbolicInt'
+    mod: 'SymbolicInt'
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature(
+            [
+                Register('u', QMontgomeryUInt(self.bitsize)),
+                Register('v', QMontgomeryUInt(self.bitsize)),
+                Register('r', QMontgomeryUInt(self.bitsize)),
+                Register('s', QMontgomeryUInt(self.bitsize)),
+                Register('m', QBit()),
+                Register('f', QBit()),
+            ]
+        )
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', u: Soquet, v: Soquet, r: Soquet, s: Soquet, m: Soquet, f: Soquet
+    ) -> Dict[str, 'SoquetT']:
+        a = bb.allocate(1)
+        b = bb.allocate(1)
+
+        v, m, f = bb.add(_KaliskiIterationStep1(self.bitsize), v=v, m=m, f=f)
+        u, v, b, a, m, f = bb.add(
+            _KaliskiIterationStep2(self.bitsize), u=u, v=v, b=b, a=a, m=m, f=f
+        )
+        u, v, b, a, m, f = bb.add(
+            _KaliskiIterationStep3(self.bitsize), u=u, v=v, b=b, a=a, m=m, f=f
+        )
+        u, v, r, s, a = bb.add(_KaliskiIterationStep4(self.bitsize), u=u, v=v, r=r, s=s, a=a)
+        u, v, r, s, b, f = bb.add(
+            _KaliskiIterationStep5(self.bitsize), u=u, v=v, r=r, s=s, b=b, f=f
+        )
+        u, v, r, s, b, a, m, f = bb.add(
+            _KaliskiIterationStep6(self.bitsize, self.mod), u=u, v=v, r=r, s=s, b=b, a=a, m=m, f=f
+        )
+
+        bb.free(a)
+        bb.free(b)
+        return {'u': u, 'v': v, 'r': r, 's': s, 'm': m, 'f': f}
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> 'BloqCountDictT':
+        return {
+            _KaliskiIterationStep1(self.bitsize): 1,
+            _KaliskiIterationStep2(self.bitsize): 1,
+            _KaliskiIterationStep3(self.bitsize): 1,
+            _KaliskiIterationStep4(self.bitsize): 1,
+            _KaliskiIterationStep5(self.bitsize): 1,
+            _KaliskiIterationStep6(self.bitsize, self.mod): 1,
+        }
+
+    def on_classical_vals(
+        self, u: int, v: int, r: int, s: int, m: int, f: int
+    ) -> Dict[str, 'ClassicalValT']:
+        a = b = 0
+        assert m == 0
+        m ^= f & (v == 0)
+        f ^= m
+
+        a ^= f & (u % 2 == 0)
+        m ^= f & (a == 0) & (v % 2 == 0)
+        b ^= a
+        b ^= m
+
+        t = (u > v) & (b == 0) & f
+        a ^= t
+        m ^= t
+
+        if a:
+            u, v = v, u
+            r, s = s, r
+
+        if f and b == 0:
+            v -= u
+            s += r
+
+        b ^= m
+        b ^= a
+        if f:
+            assert v % 2 == 0, f'{u=} {v=} {r=} {s=} {a=} {b=} {m=} {f=}'
+            v >>= 1
+        r = (r << 1) % self.mod
+        if a:
+            u, v = v, u
+            s, r = r, s
+        a ^= s == 0
+        return {'u': u, 'v': v, 'r': r, 's': s, 'a': a, 'b': b, 'm': m, 'f': f}
```

Comment:
This is a broader comment, but attaching it to these lines since this is the place where it's most acute.

Is this just a transcription of the individual bloq operations? ideally the classical implementation would be inspectably-correct so that we can write effective unit tests against it. If possible, we would like to avoid a unit test where one inscrutable routine produces the same output as another inscrutable routine (although this is better than nothing!). Some concrete suggestions and requests:

 - Can we have a unit test that proves that the bloqs in this PR are computing a modular inverse? You can take a bunch of numbers, run them through the bloq simulator, and assert that `x * (x^-1) = 1` or use `pow(x, -1, p)` to validate the output.
 - Maybe some inline comments in this `on_classical_vals`. Presumably there's some method to this madness :)

Reply:
The reason  I wrote it like this is that I didn't want to write individual tests for the `_KaliskiIterationStepX` bloqs. Instead I prefered a blackbox style test where I test the thing that we will actually use `KaliskiModInverse` which produces two register, the first has the modular inverse and the second has the ancillas. The problem here is that the value in the ancilla register is not trivial.

Reimplementing this by calling the individual _KaliskiIterationStepX.on_each is not a good idea since that will make the test useless. The test now compares the output of `_KaliskiIteration.on_classical_data` with the output of the individual steps. but if `_KaliskiIteration.on_classical_data` calls those steps then the test becomes useless.

----

> Can we have a unit test that proves that the bloqs in this PR are computing a modular inverse? 

That's what `test_kaliski_mod_inverse_classical_action` does. The modular inverse in motegomry form is such that $x_m x_m^{-1} = (x R) (x^{-1} R) R^{-1} = R = 2^{n} \mod p$ rather than $1$; it also comes from the identity $a_m \cdot b_m = (a \cdot b)_m$. I made the test clearer. 

![Screenshot from 2024-10-29 10-42-51](https://github.com/user-attachments/assets/00eb84b6-0cbe-46e1-a139-7481fa8f7f93)

> Maybe some inline comments in this on_classical_vals. Presumably there's some method to this madness

I added a docstring explaining what happens here.

Reply:
> Maybe some inline comments in this on_classical_vals. Presumably there's some method to this madness

nvm I updated the method to follow the pseudocode of the original algorithm from Fig7b in https://arxiv.org/abs/2001.09580 which is alot more readable